src/Storages/Cache/RemoteCacheController.cpp

```diff
@@ -0,0 +1,246 @@
+#include <Storages/Cache/RemoteCacheController.h>
+#include <Storages/Cache/ExternalDataSourceCache.h>
+#include <Storages/Cache/RemoteFileMetadataFactory.h>
+#include <IO/ReadBufferFromFile.h>
+#include <IO/ReadHelpers.h>
+#include <IO/ReadBuffer.h>
+#include <IO/ReadSettings.h>
+#include <Poco/JSON/JSON.h>
+#include <Poco/JSON/Parser.h>
+#include <fstream>
+
+namespace DB
+{
+namespace fs = std::filesystem;
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int LOGICAL_ERROR;
+}
+
+std::shared_ptr<RemoteCacheController> RemoteCacheController::recover(const std::filesystem::path & local_path_)
+{
+    auto * log = &Poco::Logger::get("RemoteCacheController");
+
+    if (!std::filesystem::exists(local_path_ / "data.bin"))
+    {
+        LOG_TRACE(log, "Invalid cached directory: {}", local_path_.string());
+        return nullptr;
+    }
+
+    auto cache_controller = std::make_shared<RemoteCacheController>(nullptr, local_path_, 0);
+    if (cache_controller->file_status != DOWNLOADED)
+    {
+        LOG_INFO(log, "Recover cached file failed. local path:{}", local_path_.string());
+        return nullptr;
+    }
+    try
+    {
+        cache_controller->file_metadata_ptr = RemoteFileMetadataFactory::instance().get(cache_controller->metadata_class);
+    }
+    catch (const Exception & e)
+    {
+        LOG_ERROR(log, "Get metadata class failed for {}. {}", cache_controller->metadata_class, e.message());
+        cache_controller->file_metadata_ptr = nullptr;
+    }
+    if (!cache_controller->file_metadata_ptr)
+    {
+        // do not load this invalid cached file and clear it. the clear action is in
+        // ExternalDataSourceCache::recoverCachedFilesMetadata(), because deleting directories during iteration will
+        // cause unexpected behaviors
+        LOG_ERROR(log, "Cannot create the metadata class : {}. The cached file is invalid and will be remove. path:{}",
+                cache_controller->metadata_class,
+                local_path_.string());
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Invalid metadata class:{}", cache_controller->metadata_class);
```

Comment:
log error can be removed if anyway throw.

Reply:
done